format/applebookmark/apple_bookmark.go

```diff
@@ -0,0 +1,448 @@
+	d.FieldBool("supports_rename")
+	d.FieldBool("supports_journaling")
+
+	d.FieldBool("supports_zero_runs")
+	d.FieldBool("supports_sparse_files")
+	d.FieldBool("is_journaling")
+	d.FieldBool("reserved_3")
+	d.FieldBool("supports_path_from_id")
+	d.FieldBool("supports_mandatory_byte_range_locks")
+	d.FieldBool("supports_hard_links")
+	d.FieldBool("supports_2_tb_file_size")
+
+	d.FieldRawLen("reserved_4", 3)
+	d.FieldBool("has64_bit_object_ids")
+	d.FieldBool("supports_decmp_fs_compression")
+	d.FieldBool("supports_hidden_files")
+	d.FieldBool("supports_remote_events")
+	d.FieldBool("supports_volume_sizes")
+}
+
+var cocoaTimeEpochDate = time.Date(2001, time.January, 1, 0, 0, 0, 0, time.UTC)
+
+type tocHeader struct {
+	nextTOCOffset    uint64
+	numEntries       uint64
+	entryArrayOffset int64
+}
+
+func (hdr *tocHeader) decodeEntries(d *decode.D) {
+	for k := uint64(0); k < hdr.numEntries; k++ {
+		entry := new(tocEntry)
+
+		d.FieldStruct("entry", func(d *decode.D) {
+			entry.key = d.FieldU32("key", elementTypeMap)
+
+			// if the key has the top bit set, then (key & 0x7fffffff)
+			// gives the offset of a string record.
+			if entry.key&0x80000000 != 0 {
+				d.FieldStruct("key_string", func(d *decode.D) {
+					d.SeekAbs(calcOffset(entry.key&0x7fffffff), makeDecodeRecord())
+				})
+			}
+
+			entry.recordOffset = calcOffset(d.FieldU32("offset_to_record"))
+
+			d.FieldU32("unused")
+
+			switch entry.key {
+			case elementTypeTargetFlags:
+				d.SeekAbs(entry.recordOffset, func(d *decode.D) { decodeFlagDataObject(d, decodeTgtPropertyFlagBits) })
+			case elementTypeVolumeFlags:
+				d.SeekAbs(entry.recordOffset, func(d *decode.D) { decodeFlagDataObject(d, decodeVolPropertyFlagBits) })
+			default:
+				d.SeekAbs(entry.recordOffset, makeDecodeRecord())
+			}
+
+		})
+	}
+}
+
+func decodeTOCHeader(d *decode.D) *tocHeader {
+	hdr := new(tocHeader)
+
+	d.FieldStruct("toc_header", func(d *decode.D) {
+		d.FieldU32("toc_size")
+		d.FieldU32("magic", d.AssertU(0xfffffffe))
+		d.FieldU32("identifier")
+		hdr.nextTOCOffset = d.FieldU32("next_toc_offset")
+		hdr.numEntries = d.FieldU32("num_entries_in_toc")
+		hdr.entryArrayOffset = d.Pos()
+	})
+
+	return hdr
+}
+
+type tocEntry struct {
+	key          uint64
+	recordOffset int64
+}
+
+const (
+	arrayEntrySize = 4
+	dictEntrySize  = 4
+)
+
+type posLoopDetector []int64
+
+func (pld *posLoopDetector) push(i int64, detect func()) {
+	for _, o := range *pld {
+		if i == o {
+			detect()
+		}
+	}
+	*pld = append(*pld, i)
+}
+
+func (pld *posLoopDetector) pop() {
+	*pld = (*pld)[:len(*pld)-1]
+}
+
+func (pld *posLoopDetector) pushAndPop(i int64, detect func()) func() {
+	pld.push(i, detect)
+	return pld.pop
+}
+
+func makeDecodeRecord() func(d *decode.D) {
+	var pld posLoopDetector
+
+	var decodeRecord func(d *decode.D)
+	decodeRecord = func(d *decode.D) {
+		defer pld.pushAndPop(
+			d.Pos(),
+			func() { d.Fatalf("infinite recursion detected in record decode function") },
+		)
```

Comment:
typo here, pop function is not defer called, add a `()`, after that the loop fqtest works for me 👍